src/psyclone/tests/core/symbolic_maths_test.py

```diff
@@ -0,0 +1,245 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+# Modified: A. R. Porter and R. W. Ford  STFC Daresbury Lab
+# Modified: I. Kavcic, Met Office
+
+
+''' Module containing py.test tests for dependency analysis.'''
+
+from __future__ import print_function, absolute_import
+
+import pytest
+
+from psyclone.core import SymbolicMaths
+
+
+def test_sym_maths_get():
+    '''Makes sure that the getter works as expected, especially
+    that sympy can be imported.'''
+
+    sym_maths = SymbolicMaths.get()
+    assert sym_maths is not None
+
+    assert not sym_maths.equal(None, 1)
+    assert not sym_maths.equal(2, None)
+
+
+@pytest.mark.parametrize("expressions", [(".true.", ".TRUE."),
+                                         (".false.", ".FALSE."),
+                                         ])
+def test_math_logicals(fortran_reader, expressions):
+    '''Test that the sympy based comparison handles constant numbers,
+    especially once using Fortran type specification
+    '''
+    # A dummy program to easily create the PSyIR for the
+    # expressions we need. We just take the RHS of the assignments
+    source = '''program test_prog
+                use some_mod
+                logical :: x
+                x = {0}
+                x = {1}
+                end program test_prog
+                '''.format(expressions[0], expressions[1])
+    psyir = fortran_reader.psyir_from_source(source)
+    schedule = psyir.children[0]
+
+    sym_maths = SymbolicMaths.get()
+    assert sym_maths.equal(schedule[0].rhs, schedule[1].rhs)
+
+
+@pytest.mark.parametrize("expressions", [("i", "i"),
+                                         ("2", "1+1"),
+                                         ("123_4", "123_8"),
+                                         ("123_4", "120+3"),
+                                         ("123_xx", "123"),
+                                         ("1.23E5", "123000"),
+                                         ("1.23D5", "123000"),
+                                         ("1.0E+3", "1000"),
+                                         ("1.0", "1"),
```

Comment:
FWIW:
```
    if (1 == 1.0d0) then
        print *,"EQUAL"
    endif
```
prints equal for gfortran and intel. I am pretty sure the standard specifies that the 'smaller' data type gets converted to the 'larger' (i.e. the integer becomes double - see http://rsusu1.rnd.runnet.ru/develop/fortran/prof77/node108.html - not the standard, but first useful google result) - and as such SymPy should report the same.

Reply:
I still think this is the right way, since I assume this is what a compiler would do if it evaluates constant expressions. There is a slight potential problem if you have a huge int, which gets converted to a real, and loses precision. SymPy does not lose precision. so this is the only case that I would see a problem - but I think that comparing 'small' numbers is what is happening in 99% of all cases anyway.